bevy_lint/src/lints/borrow_of_reborrowable.rs

```diff
@@ -0,0 +1,224 @@
+//! Checks for function parameters that take a mutable reference to a
+//! re-borrwable type.
+//!
+//! # Motivation
+//!
+//! Mutable references to re-borrwable types can almost always be readily
+//! converted back to an owned instance of itself, albeit with an appropriately
+//! shorter lifetime.
+//!
+//! The only time this isn't true is when the function returns referenced data
+//! that is bound to the mutable reference of the re-borrowable type.
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(&mut commands);
+//! }
+//!
+//! fn helper_function(commands: &mut Commands) {
+//!   // ...
+//! }
+//! ```
+//!
+//! Use instead:
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(commands.reborrow());
+//! }
+//!
+//! fn helper_function(mut commands: Commands) {
+//!   // ...
+//! }
+//! ```
+
+use std::ops::ControlFlow;
+
+use crate::declare_bevy_lint;
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::match_type};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl, Mutability};
+use rustc_lint::{LateContext, LateLintPass, Lint};
+use rustc_middle::{
+    ty::Interner,
+    ty::{Ty, TyKind, TypeVisitable, TypeVisitor},
+};
+use rustc_session::declare_lint_pass;
+use rustc_span::{def_id::LocalDefId, symbol::Ident, Span};
+
+declare_bevy_lint! {
+    pub BORROW_OF_COMMANDS,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Commands` or `EntityCommands` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_RESOURCE,
+    PEDANTIC,
+    "parameter takes a mutable reference to `ResMut` or `NonSendMut` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_QUERY,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Query` instead of a re-borrowed instance",
+}
+
+declare_lint_pass! {
+    BorrowOfReborrowable => [BORROW_OF_COMMANDS.lint, BORROW_OF_RESOURCE.lint, BORROW_OF_QUERY.lint]
+}
+
+impl<'tcx> LateLintPass<'tcx> for BorrowOfReborrowable {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
```

Comment:
```suggestion
    ) {
        // Closures are not currently supported, as `tcx.fn_sig()` crashes for them.
        if let FnKind::Closure = kind {
            return;
        }
```

This should fix the CI errors when running the unit tests. It should be possible to support closures in the future, but we can totally defer that for the future.

(If we do defer this, I'd like to raise an issue, add a note to the `# Known Issues` section of the module docs, and write a quick `bug_xxx.rs` UI test. I can do this, though!)